cpprevolve/revolve/gazebo/brains/DifferentialCPG.cpp

```diff
@@ -60,144 +124,871 @@ DifferentialCPG::DifferentialCPG(
+  double dt = (_time - this->previous_time);
+  this->previous_time = _time;
 
-      if (x2 == x and y2 == y and z2 == z)
+  // Perform one step
+  stepper.do_step(
+      [this](const state_type &x, state_type &dxdt, double t)
       {
-        auto input = std::get<2>(this->neurons_[{x1, y1, z1}]);
-        inputA += weightBA * input + biasA;
-      }
-    }
+        for(size_t i = 0; i < this->neurons.size(); i++){
+          dxdt[i] = 0;
+          for(size_t j = 0; j < this->neurons.size(); j++)
+          {
+            dxdt[i] += x[j]*this->ode_matrix[j][i];
+          }
+        }
+      },
+      x,
+      _time,
+      dt);
 
-    this->nextState_[i] = stateA + (inputA * _time);
-    ++i;
+  // Copy values into nextstate
+  for (size_t i = 0; i < this->neurons.size(); i++){
+    this->next_state[i] = x[i];
   }
 
-  i = 0; auto j = 0;
-  for (auto &neuron : this->neurons_)
+  // Loop over all neurons to actually update their states. Note that this is a new outer for loop
+  auto i = 0; auto j = 0;
+  for (auto &neuron : this->neurons)
   {
-    double biasA, gainA, stateA;
-    std::tie(biasA, gainA, stateA) = neuron.second;
+    // Get bias gain and state for this neuron. Note that we don't take the coordinates.
+    // However, they are implicit as their order did not change.
+    double bias, gain, state;
+    std::tie(bias, gain, state) = neuron.second;
+    double x, y, z;
+    std::tie(x, y, z) = neuron.first;
+    neuron.second = {bias, gain, this->next_state[i]};
 
-    neuron.second = {biasA, gainA, this->nextState_[i]};
-    if (i % 2 == 0)
+    // Should be one, as output should be based on +1 neurons, which are the A neurons
+    if (i % 2 == 1)
     {
-      _output[j] = this->nextState_[i];
+      // TODO: Add Milan's function here as soon as things are working a bit
+      // f(a) = (w_ao*a - bias)*gain
+
+      // Apply saturation formula
+      auto x = this->next_state[i];
+      this->output[j] = this->signal_factor*this->abs_output_bound*((2.0)/(1.0 + std::pow(2.718, -2.0*x/this->abs_output_bound)) -1);
       j++;
     }
     ++i;
   }
+
+  // Write state to file
+  std::ofstream state_file;
+  state_file.open(this->directory_name + "states.txt", std::ios::app);
+  for(size_t i = 0; i < this->neurons.size(); i++){
+    state_file << this->next_state[i] << ",";
+  }
+  state_file << "\n";
+  state_file.close();
+
+  // Write signal to file
+  std::ofstream signal_file;
+  signal_file.open(this->directory_name + "signal.txt", std::ios::app);
+  for(size_t i = 0; i < this->n_motors; i++){
+    signal_file << this->output[i] << ",";
+  }
+  signal_file << "\n";
+  signal_file.close();
+}
+
+/**
+ * Struct that holds the parameters on which BO is called. This is required
+ * by limbo.
+ */
+struct DifferentialCPG::Params{
+  struct bayes_opt_boptimizer : public limbo::defaults::bayes_opt_boptimizer {
+  };
+
+  // depending on which internal optimizer we use, we need to import different parameters
+#ifdef USE_NLOPT
```

Comment:
is this DEFINE defined? are we using nlopt?

Reply:
I'm not sure what do to with this or this comment. I've removed the second time we call:

```
#ifndef USE_NLOPT
#define USE_NLOPT //installed NLOPT
#endif

```

Reply:
I think you fixed this

Reply:
Yes, yesterday I managed to fix it :)